### DISCUSSION TITLE
How to persist ISR-cache between deploys with custom server?

### DISCUSSION THREAD

**Discussion by stnwk** [vercel/next.js#30244] [Help] [answered]
Hi :)

I have a custom server setup that is deployed on a Digital Ocean Droplet (not App Platform), so I have full file-system control.
I also have a page `item/[item_id].ts` that uses ISR to generate pages for each given `item_id` on-demand, no pages are being pre-generated during build time.

When I go on `domain.com/item/1`, the matching `item/1.html` and `item/1.json` filesare generated, written to the file-system and added to the `IncrementalCache`.

Now I make some changes in the codebase and re-deploy/re-build.

On **vercel**, the generated `item/1.html` + `item/1.json` files are being persisted across the deploys, while on my custom server the `.next` directory is always wiped clean when building

My question is: How is the behavior on vercel exactly achieved?

I know there is the config option `cleanDistDir: false`, but that would keep the `.next` directory growing forever, wouldn't it? How would I achieve this in a clean way?

Thanks for all the help :)

**Comment by stnwk** [comment]
Maybe @leerob can help?

**Reply by kawpii** [reply]
I would also like to know about this.

**Comment by stigkj** [comment]
I would think caching is done in Vercel's CDN?

**Reply by stnwk** [reply]
I'm not sure I understand this answer, how would that work?

The IncrementalStaticCache is both in-memory and on disk, meaning vercel would need to persist either memory or files on disk between deployments. Persisting the correct files on disk from `.next/cache` seems like the solution, but what files exactly so it doesn't interfere with upcoming builds?

**Comment by samcx** [comment]
Unfortunately, we can't provide code-level support for custom server setups.

As for your ISR question → https://vercel.com/docs/concepts/next.js/incremental-static-regeneration#isr:-not-just-caching.

Let us know if you have any further questions or concerns.

**Comment by MauriceArikoglu** [comment]
@stnwk you should be able to use a mounted volume to persist the next folder between redeploys. For new deploys to become effective you must use the newest build folder, while persisting the runtime generated files. For that to work you must also disable memory cache though, as only then everything is cached on filesystem. Here’s an example from the community of a similar problem specifically with kubernetes - but I think it generally also applies to your problem https://github.com/vercel/next.js/discussions/38858#discussioncomment-7461914

**Reply by leerob** [reply]
We have now the option to provide a custom cache location for ISR: https://nextjs.org/docs/app/api-reference/next-config-js/incrementalCacheHandlerPath

**Reply by MauriceArikoglu** [reply]
>  We have now the option to provide a custom cache location for ISR: https://nextjs.org/docs/app/api-reference/next-config-js/incrementalCacheHandlerPath

@leerob 

First, the documentation is not clear on supported versions for this feature.

Second, this feature is only available in app router, not the pages router. Which renders the feature unusable to all clients still using and pushing forward on the pages router.

Due to aforementioned reasons this makes a custom solution still required.

@stnwk I wouldn't mark this as answer, as it is not viable to all users.

Edit: I just saw the other comment outlining extended documentation. The above still stands for the attached link but it's nice to see custom-hosting is getting caressed.

**Reply by leerob** [reply]
Supported version are being added to the docs here https://github.com/vercel/next.js/pull/57953 as it's being stabilized. It works with both the Pages and App Router.

**Reply by MauriceArikoglu** [reply]
For everyone not wanting to sift through the changed files:
- experimental incrementalCacheHandlerPath was introduced in `v12.2`
- it is made stable and renamed to cacheHandler in `v14.1`

**Comment by leerob** [comment, marked as answer]
Hey everyone – we have just published new documentation on self-hosting, including how to configure a custom cache handler for use with ISR / Data Cache in both the Pages and App Router. These docs go in depth about how other features like Image Optimization, Middleware, Environment Variables, Build Caching, and more work when self-hosted. There's also updated Docker based examples.

Specific to this discussion, we have provided guidance on how to use runtime environment variables.

https://nextjs.org/docs/app/building-your-application/deploying
